pallets/parachain-staking/src/nominator_list.rs

```diff
@@ -0,0 +1,198 @@
+// Copyright 2019-2021 PureStake Inc.
+// This file is part of Moonbeam.
+
+// Moonbeam is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Moonbeam is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Moonbeam.  If not, see <http://www.gnu.org/licenses/>.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use crate::*;
+use sp_runtime::{
+	traits::AtLeast32BitUnsigned,
+};
+use sp_std::prelude::*;
+
+/// An in-memory struct representing a collator's list of nominators. This list is sorted by amount
+/// of stake upon every iteration and the top N nominators are trivially available at any time. In
+/// addition, several sums are maintained (also updated upon any mutation):
+///
+/// 1) the total sum of all nominators
+/// 2) the sum of the top N nominators
+/// 3) the sum of the nominators outside of the top N (the "bottom" nominators)
+///
+/// Design note: I find it to be a useful exercise to fully abstract away specifics. This first pass
+/// doesn't reflect this, but consider what this struct and its functions might look like if they
+/// had nothing to do with collators or nominators.
```

Comment:
I agree with this whole-heartedly. This data structure actually seems somewhat similar to a priority queue https://en.wikipedia.org/wiki/Priority_queue where the total stake for each candidate is the priority.